Adding Purge Instance History functionality

gled4er opened the pull request:
Hello @cgillum,

May I ask you when you have time to take a look on this pull request for the [History Retention](https://github.com/Azure/azure-functions-durable-extension/issues/17) feature?

This is the first part of the required code but I wanted to check early with you if the direction is correct.

The next steps will be to add one more API with the filter options you mentioned in the discussion for the issue and add new methods in the Durable Functions project.

Thank you!

gled4er commented:
Hello @cgillum,

I wanted to clarify with you the logic for deleting the history based on datetime period. In the case [GetOrchestrationStateAsync](https://github.com/Azure/durabletask/blob/493dc055e910a9d30f57d8f36f30fef0def7df81/src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs#L1193) we have single runtime status but on the other hand in the History table we have multiple entries for each execution that have different statuses. So I see two ways to implement the deletion of history when datetime period is provided:

1. **Support Runtime Status Filteting** - this will be the orchestration runtime status so we need first to retrieve all the orchestrations with the provided runtime status directly from the Instance Table and then query all the history entries related to them and finally delete them.
2. **Do not support Runtime Status Filtering** - do the filtering for history events only based on datetime period and then delete all of the items. I think this approach is a little bit risky because we can have orchestration that runs with huge pauses in the execution so by mistake we can delete a history item of currently running orchestration (but waiting for external event for example) that will lead to the orchestration  failure. 

For now I will proceed with option 1 but please feel free to correct my understanding and I will fix the code quickly.

Thank you!

cgillum commented on Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs:
```diff
@@ -142,6 +143,109 @@ public async Task GetAllOrchestrationStatuses()
+            {
+                string instanceId = Guid.NewGuid().ToString();
+                await host.StartAsync();
+                TestOrchestrationClient client = await host.StartOrchestrationAsync(typeof(Orchestrations.SayHelloInline), "world one", instanceId);
```
Instead of `SayHelloInline`, can we use `SayHelloActivity` (or the equivalent)? It would be good to use an orchestration that also includes history for activity functions.

gled4er replied:
Hello @cgillum,

Thank you for this comment.

This is fixed now.

Thank you!

cgillum commented on Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs:
```diff
@@ -142,6 +143,109 @@ public async Task GetAllOrchestrationStatuses()
+                await host.StopAsync();
+            }
+        }
+
```
My biggest test concern for this feature is deleting lots and lots of data in a single API call. Can we create a test that clears the history for a large number of rows in table storage? I think there could be two flavors: one which deletes a single instance with a large history (more than 100 rows in table storage) and one which has multiple instances that together have more than 100 rows in table storage.

cgillum commented on src/DurableTask.AzureStorage/AnalyticsEventSource.cs:
```diff
@@ -788,6 +790,49 @@ public void GeneralWarning(string Account, string TaskHub, string Details, strin
+        }
+
+        [Event(142, Level = EventLevel.Informational)]
+
```
Please remove the extra whitespace here.

gled4er replied:
Hello @cgillum,

This is fixed now.

Thank you for catching it.

cgillum commented on src/DurableTask.AzureStorage/Tracking/ITrackingStore.cs:
```diff
@@ -119,5 +119,20 @@ interface ITrackingStore
+        /// <param name="createdTimeFrom"></param>
+        /// <param name="createdTimeTo"></param>
+        /// <param name="runtimeStatus"></param>
+        /// <returns></returns>
```
Please fill in these documentation comments.

gled4er replied:
Hello @cgillum,

Thank you for catching that.

This is done.

Thank you!

gled4er commented:
Hello @cgillum,

Thank you for the great feedback.

I agree with everything you wrote.

I will start working on it soon and let you know for the next review round.

Thank you!

gled4er commented:
Hello @cgillum,

I wanted to confirm about the 3-rd point in your previous feedback about the permitted values for `OrchestrationStatus`. Shall we also cover `ContinuedAsNew`, `Canceled` and `Pending`? 

Thank you!

gled4er commented:
Hello @cgillum,

Shall we delete the row in the **Instances** table as well when purging the data from the **History** table and the blobs?

Thank you!

gled4er commented:
Hello @cgillum,

I worked on all issues you described in the previous comments.

May I ask you when you have time to take a look at the changes I added?

I have one pending task at the moment - **I need to research more how the blob files for large messages are created and how many intermediate blob files are produced**. During the scan of the **History** table I am able to find mainly 3 blob files - input, output from Activity Function and output from the Orchestration. But when I check the **largemessages** container where the blobs are stored I see at least 4 more blobs for the same orchestration and currently I am not sure how to get their names in order to delete them. Maybe you or @tohling can help with quick guidance on that? 

Thank you!

gled4er commented:
Hello @cgillum,

Thank you for the quick reply.

I will work on the required changes.

Thank you!

cgillum commented:
Related issue: https://github.com/Azure/azure-functions-durable-extension/issues/358

gled4er commented:
Hello @cgillum,

I worked on the last outstanding issue for this feature - **purging the large payload blobs**. 

As we discussed previously, it is a bit tricky to track all the blob names since some them disappear with the queue messages after conversion of the payload to a particular business object.  I tried multiple approaches and **decided instead of the History table to use the Instances table and store there the mapping between Instances and blob names**. I think this is better approach because the schema of the Instances tables give us easy access from any part of the code. If you think that this approach pollutes the table maybe we can create new table responsible solely for the mapping between instances and blobs. I was thinking that maybe we can use Append Blob per instance but I did not investigated fully this approach. On the other hand, if you think that we should use the **History** table for that I think we need to be very careful to avoid any possible disturbance of the replay logic. 

After trying to work with the `AzureOrchestrationService` and `MessageManager` in order to capture all the blob names I decided that the best approach will be to do it in the `ControlQueue` and `WorkItemQueue` classes. I created a new static class and method in it `LargePayloadBlobManager.AddBlobsData` that is performing batch insert to the **Instances** table. I think this is one of the advantages of the approach involving the **Instances** table or dedicated blob table. You can easily see all the changes I am describing in this comment in the last commit. 

I am very interested in your opinion!

Thank you!

gled4er commented on src/DurableTask.Core/TaskMessage.cs:
```diff
@@ -33,6 +33,12 @@ public class TaskMessage : IExtensibleDataObject
+        /// The blob name for the compressed message. This value is set if there is a compressed blob.
+        /// </summary>
+        [DataMember]
+        public string CompressedBlobName { get; set; }
```
Hello @cgillum,

Great catch! I reverted this change.

Thank you!

gled4er commented on Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs:
```diff
@@ -142,6 +143,190 @@ public async Task GetAllOrchestrationStatuses()
+                TestOrchestrationClient client = await host.StartOrchestrationAsync(typeof(Orchestrations.Factorial), 110, instanceId);
+                await client.WaitForCompletionAsync(TimeSpan.FromSeconds(30));
+                
+
```
Thank you! I will remove the extra line.

gled4er commented on Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs:
```diff
@@ -142,6 +143,190 @@ public async Task GetAllOrchestrationStatuses()
+                Assert.AreEqual(4, results.Count);
+                Assert.IsNotNull(results[0].Output.Equals("\"Done\""));
+                Assert.IsNotNull(results[1].Output.Equals("\"Done\""));
+                Assert.IsNotNull(results[2].Output.Equals("\"Done\""));
```
Great catches! 

Thank you very much!

I will optimize this code.

Thank you!

gled4er commented on Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs:
```diff
@@ -142,6 +143,190 @@ public async Task GetAllOrchestrationStatuses()
+                Assert.AreEqual(1, thirdOrchestrationStateList.Count);
+                Assert.AreEqual(thirdInstanceId, thirdOrchestrationStateList.First().OrchestrationInstance.InstanceId);
+
+                await client.PurgeInstanceHistoryByTimePeriod(startDateTime, DateTime.Now, new List<OrchestrationStatus> {OrchestrationStatus.Completed, OrchestrationStatus.Terminated, OrchestrationStatus.Failed, OrchestrationStatus.Running});
```
This is great feedback.

I will fix it.

Thank you!

gled4er commented on Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs:
```diff
@@ -142,6 +143,190 @@ public async Task GetAllOrchestrationStatuses()
+
+                firstOrchestrationStateList = await client.GetStateAsync(firstInstanceId);
+                Assert.AreEqual(1, firstOrchestrationStateList.Count);
+                Assert.IsNull(firstOrchestrationStateList.First());
```
Hello @cgillum,

The list is coming from [here](https://github.com/Azure/durabletask/blob/e272e28092209030ebb6ee37d1bd087a278621de/src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs#L420). After inspecting the code I don't think we need the list. We are attempting to get all statuses for all executions but in the **Instances** tables we have only a single line per instance. I can fix that but maybe I am missing some use case.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/AnalyticsEventSource.cs:
```diff
@@ -788,6 +790,48 @@ public void GeneralWarning(string Account, string TaskHub, string Details, strin
+        }
+
+        [Event(142, Level = EventLevel.Informational)]
+        public void PurgeInstanceHistoryTimeFilter(
```
Great optimization!

I will fix it.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -1254,6 +1254,27 @@ public async Task<string> GetOrchestrationHistoryAsync(string instanceId, string
+        /// <param name="instanceId">Instance ID of the orchestration.</param>
+        public async Task PurgeInstanceHistoryAsync(string instanceId)
+        {
+            await this.trackingStore.PurgeInstanceHistoryAsync(instanceId);
```
Thank you for this optimization.

I will fix it.

Thank you!

gled4er commented:
Hello @cgillum,

I addressed the issues you highlighted in your feedback.

I need to resolve merge conflicts and I will work on that next.

When you have time I am very interested to discuss the design of the instance to blobs mapping. I think this is the biggest item that we need to address for this pull request.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -1254,6 +1254,27 @@ public async Task<string> GetOrchestrationHistoryAsync(string instanceId, string
+        /// <param name="runtimeStatus">RuntimeStatus of orchestrations. You can specify several status.</param>
+        public async Task PurgeInstanceHistoryAsync(DateTime createdTimeFrom, DateTime? createdTimeTo, IEnumerable<OrchestrationStatus> runtimeStatus)
+        {
+            await this.trackingStore.PurgeInstanceHistoryAsync(createdTimeFrom, createdTimeTo, runtimeStatus);
```
Thank you!

I will fix it.

gled4er commented:
Hello @cgillum,

I went ahead and I added the code for using dedicated Azure Storage Table for mapping orchestration instances and large payload blobs. 

I played a little bit with the Append Blobs but there are several reasons I think the Storage Table is better approach:

- For the Append blob code we needed to make at least 2 storage request - one for checking if the Append blob already exists and one more for appending content. We also don't have batch operation since we have a blob file per orchestration. So using the Append blob will increase the requests to Azure Storage.
- The `AppendTextAsync` method does not work properly when multiple writers are calling it in parallel
- To avoid the limitation in the previous step we need to use `AppendBlockAsync` but during my testing it did not behave well most of the time
- The code for `AppendBlockAsync` required working with streams instead of strings and in my opinion the code started to look a bit complicated
- Azure Storage Emulator does not support Append Blobs and the team behind it does plan to change that. They recommend to use an open source tool but I think this will make the local dev experience complicated and we will need to add code to filter the tests targeting Append Blobs from local execution.
- On the other hand the work with table is straightforward and we are able to perform batch operations

Let me know what you think.

Thank you!

cgillum commented:
Thanks Kanio. I haven't had a chance to look at this quite yet, but I'm really eager to once I can make time (I've been busy with CRIs this past week). I'm hesitant to introduce a new table since then we have to document it and maintain it. I was hoping we could just reuse our existing tables. However, I will take a look at your code and the design constraints you mentioned to see if I can better understand where you're coming from.

gled4er commented:
Hello @cgillum,

Thank you very much for the update.

I think if we can use the **Instances** table then it will be easy. I already had a version that was doing exactly this. I will revert the latest changes so you can see my idea that is using the **Instances** table. If we want to use the History table I need to investigate further. Do you have preferences which table we should use?

Thank you!

Kanio

gled4er commented:
Hello @cgillum,

I will look more into that. I took the route of using the **Instances** table because it was straightforward but you are right that the **History** table is a better fit.

I will update you after I find out more.

Thank you!

gled4er commented:
Hello @cgillum,

I researched further the situation with adding the blob names to the **History** table. I think I have now better understanding. The problem is that currently we have the compressed blob name as property of `MessageData` object that can be accessed in `AzureStorageOrchestrationService.CompleteTaskOrchestrationWorkItemAsync` method via the session object and more specifically via its `CurrentMessageBatch` property. My idea is to try to pass the blob names to the `AzureTableTrackingStore.UpdateStateAsync` method and add them to the **History** table there. However, the main object we are passing there is `OrchestrationRuntimeState` and the best place I think we can do that is via the `HistoryEvent` (`AllEvents` or `NewEvents`)  collections. My current idea was to modify the `Input` / `Result` properties and append the blob name separated from the real content with well known marker. However, half of the events I planned to use have private setters for the `Result` property. So this approach will not work. We can pass the data as additional parameter for `AzureTableTrackingStore.UpdateStateAsync` but this will require changes in the interface in the Core project that we don't want to do. So basically my current approach is to find a way to pass the blob names to `AzureTableTrackingStore` from `AzureStorageOrchestrationService`. If you have any suggestion, please share.

Thank you!

cgillum commented:
Thanks Kanio for persisting in this. I recognize the design challenge that you mentioned. Here are my thoughts on a possible design:

1. Whenever possible, we should try to avoid making changes to DurableTask.Core. I realize having access to the `Result` property would be convenient, but I think it might be hard to justify this change. Having a immutable/read-only `Result` property is generally a good design for this framework and other parts of the code probably make certain assumptions which could be broken if we changed this.

2. The main complication right now is that we only upload inputs to blob storage deep inside of `TaskHubQueue`. I think these methods will need to return the name of the blob that was created. Maybe the methods which currently return `Task<CloudQueueMessage>` could instead return `Task<MessageData>`, and those which just return `Task` could return `Task<MessageData>`? That would give the calling code access both to `CompressedBlobName` and `OriginalQueueMessage`.

3. For `ITrackingStore.UpdateStateAsync`, maybe we could add another parameter of type `IDictionary<HistoryEvent, string>` or `IDictionary<HistoryEvent, MessageData>`. The key would be the history event with the large message. The value would be the data we need to get the blob name. When converting history events to dynamic table records, we could add the blob name to the dynamic table record we create by looking it up in this dictionary.

I think this is doable, but there may be some corner cases I'm still missing. What do you think?

gled4er commented:
Hello @cgillum,

Thank you very much for the great feedback.

I work on items 2 and 3 and I will let you know about my progress.

Thank you!

gled4er commented:
Hello @cgillum,

Thank you once more for the guidance and the directions in your previous comment!

I decided instead of modifying multiple methods in both `AzureStorageOrchestrationServiceand` and  `TaskHubQueue` to use `Dictionary<HistoryEvent, string> historyEventBlobNames` to hold all the blob names that are processed in `AzureStorageOrchestrationService` since this is the place we get all the messages from the queues. 

When I was running the tests with attribute `[DataRow(true)]` I found out that we were missing one blob name that I was able to capture only by inspecting the `session.CurrentMessageBatch`  that is why I have the code below in `AzureStorageOrchestrationService.CompleteTaskOrchestrationWorkItemAsync`:

```csharp
              if (session.CurrentMessageBatch.Any())
                {
                    foreach (MessageData messageData in session.CurrentMessageBatch)
                    {
                        if (!this.historyEventBlobNames.ContainsKey(messageData.TaskMessage.Event))
                        {
                            this.historyEventBlobNames.Add(messageData.TaskMessage.Event, messageData.CompressedBlobName);
                        }
                    }
                }
```

And then in `AzureTableTrackingStore.UpdateStateAsync` I have the following fragment that appends the blob name mapped to the `HistoryEvent` coming from the `session.CurrentMessageBatch` since I wasn't able to match it with the items in  `runtimeState.NewEvents`:

```csharp

if (historyEventBlobNames.ContainsKey(newEvents[i]))
{
       blobName = historyEventBlobNames[newEvents[i]];
       historyEventBlobNames.Remove(newEvents[i]);
}

```

I also modified a bit `AzureTableTrackingStore.CompressLargeMessageAsync` to cover the cases when we pass blob names values and handle the case when this method is called by `AzureTableTrackingStore.SetNewExecutionAsync`:

```csharp
        async Task CompressLargeMessageAsync(DynamicTableEntity entity, string previousBlobName = "",   bool isSetNewExecutionAsync = false, string exisitingPreviousBlobName = "")
        {
            string propertyKey = this.GetLargeTableEntity(entity);
            if (propertyKey != null)
            {
                string blobName = Guid.NewGuid().ToString();
                // e.g.InputBlobName, OutputBlobName, ResultBlobName
                string blobNameKey = $"{propertyKey}{BlobNamePropertySuffix}";
                byte[] messageBytes = this.GetPropertyMessageAsBytes(entity);
                await this.messageManager.CompressAndUploadAsBytesAsync(messageBytes, blobName);
                entity.Properties.Add(blobNameKey, new EntityProperty(blobName));
                if (isSetNewExecutionAsync)
                {
                    entity.Properties.Add("PreviousBlobName", new EntityProperty(
                        string.IsNullOrEmpty(exisitingPreviousBlobName) ? 
                        blobName : 
                        $"{exisitingPreviousBlobName};{blobName}"));
                }
                if (!string.IsNullOrEmpty(previousBlobName))
                {
                    entity.Properties.Add("PreviousBlobName", new EntityProperty(previousBlobName));
                }
                this.SetPropertyMessageToEmptyString(entity);
            }
        }

```

I also added a bit more code to the `AzureTableTrackingStore.SetNewExecutionAsync` to handle the case when we have multiple executions for the same orchestration:

```csharp
          TableQuery query = new TableQuery().Where(TableQuery.CombineFilters(
                TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.Equal, executionStartedEvent.OrchestrationInstance.InstanceId),
                TableOperators.And,
                TableQuery.GenerateFilterCondition("RowKey", QueryComparisons.LessThan, string.Empty))).Select(new List<string> { "PreviousBlobName"});
            TableContinuationToken token = null;
            var segment = await this.InstancesTable.ExecuteQuerySegmentedAsync(query, token);
            string previousBlobName = string.Empty;
            if(segment.Results.Any() &&
               segment.Results.First().Properties.ContainsKey("PreviousBlobName") &&
               !string.IsNullOrEmpty(segment.Results.First().Properties["PreviousBlobName"].StringValue))
            {
                previousBlobName = segment.Results.First().Properties["PreviousBlobName"].StringValue;
            }

            await this.CompressLargeMessageAsync(entity, string.Empty, true, previousBlobName);
```

I have more work on the pull request as you can see from the **TODO** comments but I wanted to check with you if the approach is good.

Please let me know what you think.

Thank you!

cgillum commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -70,6 +70,8 @@ public sealed class AzureStorageOrchestrationService :
         readonly PartitionManager<BlobLease> partitionManager;
         readonly OrchestrationSessionManager orchestrationSessionManager;
 
+        readonly IDictionary<HistoryEvent, string> historyEventBlobNames;
```
I don't recommend creating global transient state like this. It makes the code harder to reason about and can lead to hard-to-diagnose bugs and possibly memory leaks. Instead, I would much prefer local variables, parameters, and return values.

gled4er replied:
Thank you for the guidance.

I will remove that part of the code.

cgillum commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -596,6 +600,10 @@ static TaskHubInfo GetTaskHubInfo(string taskHub, int partitionCount)
                         session.TraceProcessingMessage(message, isExtendedSession: false);
+                        if (!string.IsNullOrEmpty(message.CompressedBlobName))
+                        {
+                            this.historyEventBlobNames.Add(message.TaskMessage.Event, message.CompressedBlobName);
```
I was actually expecting that all the blob name logic would be contained in CompleteTaskOrchestration... method. We would capture the blob info at the time the blob is created and persist it to the history table shortly after.

gled4er replied:
Thank you for this suggestion.

I will focus on implementing all the required logic in `AzureStorageOrchestrationService.CompleteTaskOrchestrationWorkItemAsync`.

Thank you!

gled4er commented:
Hello @cgillum,

Thank you very much for the great feedback.

Now I have better understanding about your expectations.

Thank you!

Kanio

gled4er commented on src/DurableTask.AzureStorage/MessageManager.cs:
```diff
@@ -1,4 +1,17 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//  ----------------------------------------------------------------------------------
```
Sure.

gled4er commented on src/DurableTask.AzureStorage/AnalyticsEventSource.cs:
```diff
@@ -788,6 +790,32 @@ public void GeneralWarning(string Account, string TaskHub, string Details, strin
+            string InstanceId,
+            DateTime? createdTimeFrom,
+            DateTime? createdTimeTo,
+            IEnumerable<OrchestrationStatus> runtimeStatus,
```
Thank you for catching that!

I will fix it.

gled4er commented:
Hello @cgillum,

Thank you for the great feedback. 

I actually tried to reuse the `InputBlob` field for the `MessadaData` data payload serialized as blob in `MessageManager.SerializeMessageDataAsync` but I got an error pointing to non-deterministic workflow detected. I think it was happening because during replay the runtime was expecting for find `DynamicTableEntity` serialized but actually the content was `MessageData` inside. I will try to debug more and give you the actual method that are first detecting the error. This week we have Serverless Openhack so maybe I will be a bit slower.

The only other point I wanted to make is for the `IntialInputBlobName`. I use it  in `AzureTableTrackingStore.CompressLargeMessageAsync` when I save the initial input blob.  

Next, when I attempt to delete the blob name stored in `IntialInputBlobName` I access it 
`AzureTableTrackingStore.DeleteAllDataForOrchestrationInstance` but via the object property `OrchestrationInstanceStatus.InitialInputBlobName`:

```csharp
                if (!string.IsNullOrEmpty(orchestrationInstanceStatus.InitialInputBlobName))
                {
                    blobDeleteTaskList.Add(this.messageManager.DeleteBlobAsync(orchestrationInstanceStatus.InitialInputBlobName));
                }

```

**All the feedback you gave me is great and I will work on fixing all the issues as soon as possible!**

**Thank you!**

gled4er commented on src/DurableTask.AzureStorage/AnalyticsEventSource.cs:
```diff
@@ -788,6 +791,34 @@ public void GeneralWarning(string Account, string TaskHub, string Details, strin
+                createdTimeFrom ?? new DateTime(1900, 1, 1),
+                createdTimeTo ?? new DateTime(1900, 1, 1),
+                runtimeStatus != null ? 
+                string.Join(";", runtimeStatus.Select(x=>x.ToString()).ToArray()) : 
```
Hello @cgillum,

Thank you very much for explaining more about ETW manifest.

I will fix the both issues.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/MessageManager.cs:
```diff
@@ -175,5 +194,11 @@ internal async Task UploadToBlobAsync(byte[] data, int dataByteCount, string blo
+        internal async Task DeleteBlobAsync(string blobName)
+        {
+            CloudBlockBlob cloudBlockBlob = this.cloudBlobContainer.GetBlockBlobReference(blobName);
+            await cloudBlockBlob.DeleteIfExistsAsync();
```
Great feedback! Thanks to your comments I will use this pattern more often. It showed during this pull request on multiple times.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Messaging/TaskHubQueue.cs:
```diff
@@ -64,22 +64,35 @@ abstract class TaskHubQueue
             try
             {
+                Tuple<MessageData, CloudQueueMessage> tuple = await this.CreateOutboundQueueMessageAsync(sourceInstance, this.storageQueue.Name, message);
                 await this.storageQueue.AddMessageAsync(
```
This is great feedback!

I will fix it and I will stick to this approach in future.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -318,7 +339,7 @@ public override async Task<IList<string>> RewindHistoryAsync(string instanceId,
             var orchestratorStartedFilterCondition = new StringBuilder(200);
 
-            orchestratorStartedFilterCondition.Append("PartitionKey eq ").Append(Quote).Append(instanceId).Append(Quote); // = TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.Equal, instanceId);
+            orchestratorStartedFilterCondition.Append($"{PartitionKeyProperty} eq ").Append(Quote).Append(instanceId).Append(Quote); // = TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.Equal, instanceId);
```
Thank you @cgillum!

I will use this approach in future.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -331,7 +352,7 @@ public override async Task<IList<string>> RewindHistoryAsync(string instanceId,
             var rowsToUpdateFilterCondition = new StringBuilder(200);
 
-            rowsToUpdateFilterCondition.Append("PartitionKey eq ").Append(Quote).Append(instanceId).Append(Quote);
+            rowsToUpdateFilterCondition.Append($"{PartitionKeyProperty} eq ").Append(Quote).Append(instanceId).Append(Quote);
```
Thank you!

I will fix that!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -380,7 +401,7 @@ public override async Task<IList<string>> RewindHistoryAsync(string instanceId,
                     var soFilterCondition = new StringBuilder(200);
 
-                    soFilterCondition.Append("PartitionKey eq ").Append(Quote).Append(instanceId).Append(Quote);
+                    soFilterCondition.Append($"{PartitionKeyProperty} eq ").Append(Quote).Append(instanceId).Append(Quote);
```
Thank you!

I will fix that!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -516,18 +539,172 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+                }
+            }
+
+            this.stats.TableEntitiesRead.Increment(tableEntitiesRead);
```
This is great approach!

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -516,18 +539,172 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+                {
+                    batch.Delete(itemForDeletion);
+                    if (itemForDeletion.Properties.ContainsKey($"{InputProperty}{BlobNamePropertySuffix}") &&
+                        !string.IsNullOrEmpty(itemForDeletion.Properties[$"{InputProperty}{BlobNamePropertySuffix}"].StringValue))
```
Thank you! 

I will fix that. Now after understanding better the performance impact, this is natural choice.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -860,18 +1051,25 @@ void SetTablePropertyForMessage(DynamicTableEntity entity, DynamicTableEntity or
         }
 
-        async Task CompressLargeMessageAsync(DynamicTableEntity entity)
+        async Task CompressLargeMessageAsync(DynamicTableEntity entity, string messageDataBlobName = "", bool isSetNewExecutionAsync = false)
```
Great!

I will update it!

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Utils.cs:
```diff
@@ -50,9 +50,27 @@ public static async Task ParallelForEachAsync<T>(this IReadOnlyList<T> items, in
+                    messageDataList.Add(await task);    
+                }
             }
+            return messageDataList;
```
This is great! I missed this overload.

Thank you very much!

gled4er commented on src/DurableTask.AzureStorage/Utils.cs:
```diff
@@ -50,9 +50,27 @@ public static async Task ParallelForEachAsync<T>(this IReadOnlyList<T> items, in
+            }
+        }
 
+        public static async Task<List<TP>> ParallelForEachAsync<T,TP>(this IReadOnlyList<T> items, int maxConcurrency, Func<T, Task<TP>> action)
```
Thank you!

I will fix it.

gled4er commented on src/DurableTask.AzureStorage/Utils.cs:
```diff
@@ -67,5 +85,18 @@ static async Task InvokeThrottledAction<T>(T item, Func<T, Task> action, Semapho
             }
         }
+
+        static async Task<TP> InvokeThrottledAction<T,TP>(T item, Func<T, Task<TP>> action, SemaphoreSlim semaphore)
```
I love the consistency with .NET Framework.

I will optimize the code!

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Utils.cs:
```diff
@@ -50,9 +50,27 @@ public static async Task ParallelForEachAsync<T>(this IReadOnlyList<T> items, in
 
+        public static async Task<List<TP>> ParallelForEachAsync<T,TP>(this IReadOnlyList<T> items, int maxConcurrency, Func<T, Task<TP>> action)
+        {
+            var messageDataList = new List<TP>();
```
This is great and will keep the nature of the method generic!

Thank you!

gled4er commented on src/DurableTask.AzureStorage/AnalyticsEventSource.cs:
```diff
@@ -788,6 +791,34 @@ public void GeneralWarning(string Account, string TaskHub, string Details, strin
+            string TaskHub,
+            string InstanceId,
+            DateTime? createdTimeFrom,
+            DateTime? createdTimeTo,
```
Thank you very much!

I will fix it.

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -516,18 +539,172 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+        /// <inheritdoc />
+        public override  async Task PurgeInstanceHistoryAsync(DateTime createdTimeFrom, DateTime? createdTimeTo, IEnumerable<OrchestrationStatus> runtimeStatus)
+        {
+            var stopwatch = new Stopwatch();
```
Yes, thank you very much for catching it!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -516,18 +539,172 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+                    $"{InputProperty}{BlobNamePropertySuffix}",
+                    $"{ResultProperty}{BlobNamePropertySuffix}",
+                    MessageDataBlobNameProperty
+                });
```
Great!

I will fix it!

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -516,18 +539,172 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+                pageOffset += batchForDeletion.Count;
+            }
+
+            await Task.WhenAll(blobDeleteTaskList);
```
**This comment is a great example why I love to work on Durable Functions! Thinking for the most reliable design and always be careful for the performance impact give us invaluable experience as developers and help us grow with every new commit!**

**Thank you for the opportunity!**

I will fix it!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -606,7 +782,8 @@ public override Task StartAsync()
             string executionId,
-            string eTagValue)
+            string eTagValue,
+            IDictionary<HistoryEvent, string> historyEventBlobNames = null)
```
Thank you!

I will fix it.

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -626,13 +803,27 @@ public override Task StartAsync()
 
-                await this.CompressLargeMessageAsync(entity);
+                string blobName = string.Empty;
+                if (historyEventBlobNames.ContainsKey(newEvents[i]))
```
Thank you!

I will fix it and I will use this approach from now on.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -626,13 +803,27 @@ public override Task StartAsync()
+                if (i == newEvents.Count - 1 && historyEventBlobNames.Keys.Count > 0)
+                {
+                    HistoryEvent key = historyEventBlobNames.Keys.First();
+                    entity.Properties[MessageDataBlobNameProperty] = new EntityProperty(historyEventBlobNames[key]);
```
This is great!

I will improve!

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -38,8 +38,13 @@ class AzureTableTrackingStore : TrackingStoreBase
+        const string RowKeyProperty = "RowKey";
+        const string PartitionKeyProperty = "PartitionKey";
         const string BlobNamePropertySuffix = "BlobName";
+        const string MessageDataBlobNameProperty = "MessageDataBlobName";
```
Thank you for the comment!

I changed the value and the name of this property to be:

```csharp
         const string QueueInputBlobNameProperty = "QueueInputBlobName";
```

I will investigate further but based on my current understanding I think we should have two separate entities - one for `InputBlobName` holding the name of blob with content serialized `DynamicTableEntity` coming from `AzureTableTrackingStore.CompressLargeMessageAsync` and  another one `QueueInputBlobName` holding names of previously created blobs in `MessageManager.SerializeMessageDataAsync` whose content is serialized `MessageData` object. 

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -860,18 +1051,25 @@ void SetTablePropertyForMessage(DynamicTableEntity entity, DynamicTableEntity or
                 entity.Properties.Add(blobNameKey, new EntityProperty(blobName));
+                if (isSetNewExecutionAsync)
+                {
+                    entity.Properties.Add(IntialInputBlobNameProperty, new EntityProperty(blobName));
```
Hello @cgillum,

Based on my testing, we always update the row in **Instances** table and we overwrite the `InputBlobName` value with new one. This is ok for all changes but the first since all the updates but the first and initial `InputBlobName` value are also stored in the **History** table so we can get them from there. And I need the `InitialInputBlobName` to store the very first input blob name. Otherwise, we will lose this value and we will not delete the blob. I was also thinking for reusing the `InputBlobName` and append values over there in the format **blob-name-1;blob-name2;blob-name3;** but this will affect the performance since we need to read the table row first, get the value and append new input blob name compared to elegant 1-request **InsertOrMerge** operation we are using now. Also I don't like this approach since the list can grow with time. 

Let me know what you think.

Thank you!

gled4er commented:
Hello @cgillum,

I worked on all the issues you highlighted. **I think the main item we need to align our understanding is for the  additional properties I added to the **History** and **Instances** table.** I still think we need them but as I mentioned earlier I will double check and update you.

Thank you!

gled4er commented:
Hello @cgillum,

I was able to remove the additional column I added to the schema of the **History** table.

Now I try to think how to remove the one I added to the **Instances** table.

Thank you!

gled4er commented:
Hello @cgillum,

I removed the second column I added to the schema of the **Instances** table. With that I think we resolved all the major issues in this PR.

May I ask you to take a look at the latest changes and let me know what should be fixed next?

Thank you!

cgillum commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -860,18 +1063,50 @@ void SetTablePropertyForMessage(DynamicTableEntity entity, DynamicTableEntity or
+                !string.IsNullOrEmpty(inputBlobNameProperty.StringValue);
+        }
+
+        async Task CompressLargeMessageAsync(DynamicTableEntity entity, string messageDataBlobName)
```
Why do we need the `messageDataBlobName` parameter? There is already a `blobName` local variable in this method which contains the name of the blob. What is the difference between `blobName` and `messageBlobName`? Intuitively it seems to me that we only need `blobName` and that each `DynamicTableEntity` should have no more than one blob reference.

gled4er replied:
Thank you for the comment.

For each history event we can have up to 2 blob files. The first one is created in [SerializeMessageDataAsync](https://github.com/Azure/durabletask/blob/80ddcda54e8960779352640c88878df8656733ba/src/DurableTask.AzureStorage/MessageManager.cs#L38) when `MessageData` object associated with the queue message for that event has large size. We use this blob temporary and when we deserialize the queue message data in [DeserializeQueueMessageAsync](https://github.com/Azure/durabletask/blob/80ddcda54e8960779352640c88878df8656733ba/src/DurableTask.AzureStorage/MessageManager.cs#L66) we lose it. With the latest changes we made we track these blobs and pass them to [UpdateStateAsync](https://github.com/Azure/durabletask/blob/80ddcda54e8960779352640c88878df8656733ba/src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs#L605) via new parameter `IDictionary<HistoryEvent, string> outputBlobNames`. Inside `UpdateStateAsync` we check if the current history every is associated with any of the blob names in `IDictionary<HistoryEvent, string> outputBlobNames` if so we pass it to the `CompressLargeMessageAsync` via the `messageDataBlobName` parameter. Inside the `CompressLargeMessageAsync`  we create also new blob but this time based on the DynamicTableEntity for this history event. This is in contrast to the blob name passed via `messageDataBlobName` parameter which stores serialized `MessageData` object. 

Let me know if this makes it clear or not.

Thank you!

cgillum commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -626,13 +802,35 @@ public override Task StartAsync()
+                    outputBlobNames.Remove(newEvents[i]);
+                }
+
+                await this.CompressLargeMessageAsync(entity, blobName);
```
Related to my other comment, I don't understand why we would pass `blobName` as a parameter here. It seems to me that we've already established that this entity needs to include the name of a blob. Why can't we simply update the entity inside the current method?

I'm thinking something like this:
```csharp
if (outputBlobNames.TryGetValue(historyEvent, out string blobName)
{
    string propertyName = 
        entity.Properties.ContainsKey(InputProperty) ? InputBlobNameProperty :
        entity.Properties.ContainsKey(ResultProperty) ? ResultBlobNameProperty :
        entity.Properties.ContainsKey(OutputProperty) ? OutputBlobNameProperty : null;
    if (propertyName != null)
    {
        entity.Properties[propertyName] = blobName;
    }

    outputBlobNames.Remove(historyEvent);
}
```
In reality, I expect this only applies to `InputProperty` since the others (`ResultProperty` and `OutputProperty`) should be getting handled already by `CompressLargeMessageAsync`. If we do this, then we don't need the logic below here either, right?

gled4er replied:
Thank you!

This is great optimization!  I modified a bit the code you provided. The idea is to use one of the unused properties to hold the `MessageData` blob name with the following variation of your code:

```csharp

await this.CompressLargeMessageAsync(entity);

 if (outputBlobNames.TryGetValue(historyEvent, out string messageDataInputBlobName))
                {
                    string propertyName =
                            entity.Properties.ContainsKey(InputProperty) ? ResultBlobNameProperty :
                            entity.Properties.ContainsKey(ResultProperty) ? InputBlobNameProperty :
                            null;
                    if (propertyName != null)
                    {
                        entity.Properties[propertyName] = new EntityProperty(messageDataInputBlobName);
                    }
                    outputBlobNames.Remove(historyEvent);
                }
```
If we already have set `InputProperty` and then in `CompressLargeMessageAsync` we also set `InputBlobNameProperty` , then let's use `ResultBlobNameProperty` to store the `MessageData` blob name and when `ResultProperty` is set and in `CompressLargeMessageAsync` we also set `ResultBlobNameProperty`, we will use `InputBlobNameProperty` for the `MessageData` blob name  based on the fact that the history event types does not set both `InputProperty` and `ResultProperty` at the same time.

 Let me know if this is acceptable approach.

Thank you!

cgillum commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -860,18 +1063,50 @@ void SetTablePropertyForMessage(DynamicTableEntity entity, DynamicTableEntity or
                 byte[] messageBytes = this.GetPropertyMessageAsBytes(entity);
                 await this.messageManager.CompressAndUploadAsBytesAsync(messageBytes, blobName);
                 entity.Properties.Add(blobNameKey, new EntityProperty(blobName));
+                if (!string.IsNullOrEmpty(messageDataBlobName))
```
It seems to me that `propertyKey` would always be null for the case that actually matters, right? I think the entire problem we're trying to work around with your PR is that `entity` doesn't contain a large message field for function inputs, so this method doesn't know it needs to do anything. If that is the case, this code isn't actually doing anything that hasn't already been done.

gled4er replied:
Thank you for the comment! 

The method was simplified in a previous commit so I don't think this is now applicable.

Thank you!

gled4er commented:
Hello @cgillum,

Thank you for the great feedback.

I will work on fixing the issues and I will update you.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -516,18 +539,171 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+
+                storageRequests++;
+                tableEntitiesRead += orchestrationStates.Count - previousCount;
+                this.stats.TableEntitiesRead.Increment(tableEntitiesRead);
```
Hello @cgillum,

Thank you for this clarification. Your suggestion is great and it will simplify the code!

I will fix it.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -516,18 +539,171 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+            Stopwatch stopwatch = Stopwatch.StartNew();
+
+            TableQuerySegment<OrchestrationInstanceStatus> segment = 
+                await this.InstancesTable.ExecuteQuerySegmentedAsync(query, null);
```
Thank you!

I will fix it.

cgillum replied:
Just a reminder that we still need to update the `stats` here.

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -516,18 +539,171 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+                await Task.WhenAll(blobDeleteTaskList);
+                storageRequests += blobDeleteTaskList.Count;
+                await this.HistoryTable.ExecuteBatchAsync(batch);
+                storageRequests++;
```
Thank you for catching that. I agree for the `this.stats.StorageRequests` however the batch contains only delete operations so I thought it will be better to count it as additional storage request than `this.stats.TableEntitiesWritten` since we delete the rows. Please correct me if my understanding is not complete. How do we count the delete operations via `this.stats`?

Thank you!

cgillum replied:
`TableEntitiesWritten` is basically supposed to track any operations against table storage which are not read operations, so it is meant to include inserts, updates, and deletes.

gled4er replied:
Got it. Thank you!

Now it is clear. I was confused by the names and the usage. I saw it mainly used for inserts, upserts and merges.

I will fix it.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -516,18 +539,171 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+                RowKey = string.Empty,
+                ETag = "*"
+            }));
+            storageRequests++;
```
Thank you!
I will increment `this.stats.StorageRequests` for now.
Thank you!

cgillum replied:
I would prefer we call `this.stats.StorageRequests.Increment()` here. It's easier to keep track from a code correctness perspective if we make the increment calls right next to the actual I/O operation. The overhead is negligible.

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -516,18 +539,171 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+                }
+
+                await Task.WhenAll(blobDeleteTaskList);
+                storageRequests += blobDeleteTaskList.Count;
```
Thank you!

I will do it.

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -658,7 +856,12 @@ public override Task StartAsync()
                         orchestrationInstanceUpdate.Properties["CreatedTime"] = new EntityProperty(executionStartedEvent.Timestamp);
                         orchestrationInstanceUpdate.Properties["RuntimeStatus"] = new EntityProperty(OrchestrationStatus.Running.ToString());
-                        this.SetTablePropertyForMessage(entity, orchestrationInstanceUpdate, InputProperty, InputProperty, executionStartedEvent.Input);
+                        if (!(await this.IsInputBlobNameAlreadySet(orchestrationInstanceUpdate.PartitionKey)))
```
Thank you for the comment.

Based on my testing I see that without the check I added we will lose track of one input blob file for the row in the **Instances** table and then it will stay undeleted. 

The chain of events in `AzureTableTrackingStore` is as follows:

1. `SetNewExecutionAsync` is called and new input blob file is generated - `f5daf826-f40b-4e43-81dc-4d086603521f`
2. We reached the following line:
```csharp
this.SetTablePropertyForMessage(entity, orchestrationInstanceUpdate, InputProperty, InputProperty, executionStartedEvent.Input);
```

The value of the `InputBlobName` for the row in the **Instances** table is modified with the name of the input blob created for the history event to be inserted in the **History** table - `4132cc9e-3af0-42b0-9960-96518d185b88` and if we execute the line above without any check we lose the original value `f5daf826-f40b-4e43-81dc-4d086603521f` and later we can't delete the blob.

So the logic I added is preventing the modification the `InputBlobName` value in the **Instances** table if it is already set in order to keep the initial value and later be able to delete the blob.

Maybe there is better approach, if so please share.

Thank you!

cgillum replied:
Got it, thank you for clarifying the chain of events. It's really helpful, and it makes me wonder this: if `SetNewExecutionAsync` has already written the input blob for the orchestration, then can we simply remove the `SetTablePropertyForMessage` call? Correct me if I'm wrong, but it seems to be a redundant update.

Actually, in the ContinueAsNew case, I assume we also come down this code path and might have a large blob to set as the new input. In that case, I think we *do* want to set a new input here. So that makes me think of a different question: do we really need to persist the blob name in the **Instances** table at all? Is it not already persisted in the **History** table? Intuitively, I feel like we don't need a blob name in the **Instances** table.  Here is the sequence of events in my head:

**New instance:**
1. New instance is started by the client. A large blob is created and the ExecutionStarted message gets added to the control queue, pointing to the large blob.
2. `SetNewExecutionAsync` gets called, but does *not* track the name of the blob. The blob name is only known by the queue message.
3. The orchestration starts and awaits. `CompleteTaskOrchestrationAsync` is called, which calls `UpdateStateAsync`. My hope/expectation is that your new `outputBlobNames` dictionary parameter contains an entry for `ExecutionStarted` --> blobName. We then update just the **History** table with this blob name.
4. After the instance completes, a scheduled purge operation simply looks at the history to delete all blobs.

**ContinueAsNew**
1. Orchestration calls the `ContinueAsNew` API to start a new generation.
2. The orchestration completes and `CompleteTaskOrchestrationAsync` is called. The `ExecutionStarted/ContinuedAsNew` message gets added to the control queue with a pointer to a new input blob. **In this case, we probably want to overwrite the existing input blob**. I'm not sure if we have enough data to know the old blob name though. In that case, we will need to come up with a predictive name for all orchestration input blobs instead of using `Guid.NewGuid`. Maybe it can contain the instance ID.
3. When the new execution starts and awaits, we do exactly as we did for (3) above.
4. After the instance completes, we do the same thing as (4) above.

I think this might be a clean model and reduces the number of storage requests since we don't have to check for the existence of the blob. Do you think it is possible based on your understanding of the code?

gled4er replied:
Hello @cgillum,

Thank you for the new ideas!

I think they sound very good.

I will test them and let you know soon what is the progress.

Thank you!

gled4er replied:
Hello @cgillum,

I started to work on the ideas you shared. Actually we have one blob that is not included in the History table. If we execute `LargeTextMessagePayloads` integration test we end up with 7 unique blobs - 1 in the **Instances** table (**9e1dc0e6-6f80-4bd0-bca1-1743357c9be2**) and other 6 in the **History** table (**5520d1b5-6df4-4d71-80b9-1be3dff5f566, 5349abc2-e1ff-406a-bbda-8facb27a8afd, a4a71ca0-8ce5-4d6d-93a0-66990712b750, 0dc5a874-d6c2-498d-a632-fbc7707c071e, 2dcfaefb-20f8-4999-bdb3-ce96857b4e8a, 2dbd4dc9-27d3-4629-ac09-c68ef143d61f**). We have overlap between the **Instances** table and **History** table for the Output blob (2dbd4dc9-27d3-4629-ac09-c68ef143d61f). However, if we decide to ignore the blobs in the **Instances** table we will lose track of (**9e1dc0e6-6f80-4bd0-bca1-1743357c9be2**). I will check the code and see if we can get it somehow in the **History** table and update you here.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -516,18 +539,171 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+                int previousCount = orchestrationStates.Count;
+                storageRequests++;
+                tableEntitiesRead += segment.Results.Count;
+                this.stats.TableEntitiesRead.Increment(tableEntitiesRead);
```
Thank you!

I will fix it!

gled4er commented:
Hello @cgillum,

I added the code for removing the `InputBlobName` and `OutputBlobName` from the **Instances** table however, I think if we go with this approach maybe it will be a good idea to have a two API-s for checking the status of an orchestration - one for getting the `RuntimeStatus` and another one for getting the `Input` and `Output` so we can limit the calls to the **History** table. 

Let me know what you think. 

Thank you!

cgillum commented:
Oh, that's probably a good idea. It could dramatically simplify your work if we just have to delete a container or subdirectory.

If we do this, then we just have to make sure it is not a breaking change for existing instances. Otherwise we would have to wait to release this.

gled4er commented:
Hello @cgillum,

In the last commit I implemented the idea of using a dedicated Azure Storage blob container per orchestration with the name **largemesagges-instanceId**. I am very excited about this approach because I will be able to rollback all of the changes I made for tracking the all the blobs used by the runtime. I started removing part of the code and tomorrow (JST) I will continue. I am pretty confident that we will be able to reduce the size of the code this PR introduces and more importantly we will not need the changes I introduced to the Core project. The targeted tests on purging and large text input are passing and tomorrow I will run the rest as well.  I still have work to complete the changes but I think this should be lightest approach in terms of new code and Azure Storage performance. 

Please let me know if you have any concerns before I continue with removing part of my past code. 

The task I see before completing this feature are:

1. **Finish the testing of the new container per instances approach**
2. **Remove all unused code that was added previously**
3. **Test and work on the Continue As New scenario**
4. **Create PR for Durable Functions including integration code for calling the new methods from Durable Task code**
5. **Documentation of the new public API in Durable Functions project**

My goal is to complete all tasks up to 3 tomorrow (JST) and try also 4 based on the instructions Brandon gave me.

Thank you!

cgillum commented:
@gled4er I'm happy to hear that (but so sorry that you have been working so late)!

One thought: I'm not sure what the overhead of creating storage containers is. As an alternate to creating a container per instance, could we prefix the name of the blobs with the instance ID, like `/{container}/{instance-id}/{blob-id}.json.gz`?  The delete won't be as efficient since we have to do a ListBlobs operation to find all the blobs for deleting, but it might be safer from an Azure Storage usage perspective. I think it's still in the same spirit as your current design though. Let me know what you think.

gled4er commented:
Hello @cgillum,

Thank you for the optimization. I will follow your advice. 

It is always pleasure to work on Durable Functions so please don't worry about it.

I will update you when the required changes are committed.

Thank you!

gled4er commented:
Hello @cgillum,

I wanted to share an update for the work on this PR. Please find the details below:

1. Finish the testing of the new container per instances approach - **Completed**
2. Remove all unused code that was added previously - **Completed** 
3. Test and work on the Continue As New scenario - **Completed** 
4. Create PR for Durable Functions including integration code for calling the new methods from Durable Task code - **To Be Done**
5. Documentation of the new public API in Durable Functions project - **To Be Done**

May I ask you when you have time to check the latest changes and let me know what should be fixed next?

Thank you for your support!

cgillum commented on src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs:
```diff
@@ -763,10 +762,12 @@ bool IsExecutableInstance(OrchestrationRuntimeState runtimeState, IList<TaskMess
             string instanceId = workItem.InstanceId;
             string executionId = runtimeState.OrchestrationInstance.ExecutionId;
 
+            var outputBlobNames = new Dictionary<HistoryEvent, string>();
```
This is no longer needed, right?

gled4er replied:
Thank you for catching that!

I will fix it.

Thank you!

cgillum commented on src/DurableTask.AzureStorage/MessageManager.cs:
```diff
@@ -35,7 +56,10 @@ public MessageManager(CloudBlobClient cloudBlobClient, string blobContainerName)
-        public async Task<string> SerializeMessageDataAsync(MessageData messageData)
+        /// <param name="messageData">Instance of <see cref="MessageData"/></param>
+        /// <param name="instanceId">Instance ID</param>
+        /// <returns>Instance of Tuple mapping instance of <see cref="MessageData"/> and blob name</returns>
```
Looks like we're returning `string` instead of `Tuple`. Does this documentation comment need to be updated?

gled4er replied:
Yes, thank you for catching that.

I will fix it.

cgillum commented on src/DurableTask.AzureStorage/MessageManager.cs:
```diff
@@ -172,8 +195,52 @@ internal MessageFormatFlags GetMessageFormatFlags(MessageData messageData)
         {
             await this.cloudBlobContainer.CreateIfNotExistsAsync();
-            CloudBlockBlob cloudBlockBlob = this.cloudBlobContainer.GetBlockBlobReference(blobName);
+            CloudBlockBlob cloudBlockBlob = cloudBlobContainer.GetBlockBlobReference(blobName);
```
Can you add back the `this.` prefix here?

gled4er replied:
Thank you!

I will.

cgillum commented on src/DurableTask.AzureStorage/MessageManager.cs:
```diff
@@ -172,8 +195,52 @@ internal MessageFormatFlags GetMessageFormatFlags(MessageData messageData)
             await cloudBlockBlob.UploadFromByteArrayAsync(data, 0, dataByteCount);
         }
+
+        internal string GetNewLargeMessageBlobName(string instanceId)
```
Add a comment that shows an example of a blob path.

gled4er replied:
Great idea!

I will add the comment.

Thank you!

cgillum commented on src/DurableTask.AzureStorage/MessageManager.cs:
```diff
@@ -172,8 +195,52 @@ internal MessageFormatFlags GetMessageFormatFlags(MessageData messageData)
+                Append(instanceId).
+                Append(LargeMessageBlobNameSeparator).
+                Append(Guid.NewGuid().ToString().
+                ToLowerInvariant()).
```
`ToLowerInvariant` is not needed here since `Guid.NewGuid().ToString()` is already lowercase.

cgillum commented on src/DurableTask.AzureStorage/Messaging/ControlQueue.cs:
```diff
@@ -89,7 +89,6 @@ await batch.ParallelForEachAsync(async delegate (CloudQueueMessage queueMessage)
                         });
 
                         this.backoffHelper.Reset();
-
```
Can this change be undone?

gled4er replied:
Sure. 

I will add the a new line here.

Thank you!

cgillum commented on src/DurableTask.AzureStorage/MessageManager.cs:
```diff
@@ -35,7 +56,10 @@ public MessageManager(CloudBlobClient cloudBlobClient, string blobContainerName)
+        /// <param name="messageData">Instance of <see cref="MessageData"/></param>
+        /// <param name="instanceId">Instance ID</param>
+        /// <returns>Instance of Tuple mapping instance of <see cref="MessageData"/> and blob name</returns>
+        public async Task<string> SerializeMessageDataAsync(MessageData messageData, string instanceId)
```
Can't `instanceId` be inferred from `messageData` by using `messageData.TaskMessage.OrchestrationInstance.InstanceId`? It seems we don't need this new parameter.

gled4er replied:
Great optimization!

Thank you for catching it!

cgillum commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -39,8 +39,16 @@ class AzureTableTrackingStore : TrackingStoreBase
+        const string EventTypeProperty = "EventType";
         const string SentinelRowKey = "sentinel";
+        const char Quote = '\'';
+        const string InstnacesTableInformationMessageForLargeDataBlobs = "Too large to display. Please check History table for the actual data.";
```
Typo: Instnaces --> Instances

gled4er replied:
Thank you!

I will fix it.

cgillum commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -301,7 +323,9 @@ private async Task<List<DynamicTableEntity>> QueryHistoryForRewind(string filter
-                entities[0].Properties["ExecutionId"].StringValue,
+                entities.Count > 0 && entities.First().Properties.ContainsKey("ExecutionId") ?
+                entities[0].Properties["ExecutionId"].StringValue :
+                string.Empty,
```
Instead of having this complex expression inline in the method call, let's do it outside and save the result into a local variable named `executionId`.

gled4er replied:
Great code readability improvement!

Thank you!

cgillum commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -565,18 +591,148 @@ private async Task<IList<OrchestrationState>> QueryStateAsync(TableQuery<Orchest
+                }
+
+                int requests = await this.messageManager.DeleteLargeMessageBlobs(orchestrationInstanceStatus.PartitionKey);
+                this.stats.StorageRequests.Increment(requests);
```
In general, `stats` should only be updated next to the code which actually does the storage operation. In this case, I would expect the `Increment` call to happen inside the implementation of `DeleteLargeMessageBlobs`, not here.

gled4er replied:
Thank you for your comment. I decided to pass `AzureStorageOrchestrationServiceStats` stats as parameter for `DeleteLargeMessageBlobs`. I was thinking first to add it as dependency via the `MessageManager` constructor but unfortunately `AzureStorageOrchestrationServiceStats` is less visible than `MessageManager`. Another option was to move `DeleteLargeMessageBlobs` in `AzureTableTrackingStore` but I think logically the method is more related to the exchanged queue messages. Moreover, in `AzureTableTrackingStore` we don't have `CloudBlobContainer` dependency. 

Please let me know if another approach will be better.

Thank you!

cgillum commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -1016,22 +1174,59 @@ async Task CompressLargeMessageAsync(DynamicTableEntity entity)
+                InputBlobNameProperty);
+        }
+
+        private async Task<string> GetPropertyValue(
```
Methods that do async work should have their names end in `Async`.

gled4er replied:
Thank you I will fix the issues you highlighted above.

cgillum commented on src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:
```diff
@@ -1016,22 +1174,59 @@ async Task CompressLargeMessageAsync(DynamicTableEntity entity)
+            queryBuilder.Append(" and ").Append(EventTypeProperty).Append(" eq ").Append(Quote).Append(eventType.ToString()).Append(Quote);
+
+            CancellationToken cancellationToken;
+            List<DynamicTableEntity> entitiesToCheck = await this.QueryHistoryForRewind(
```
Why is this method calling into Rewind code? I would expect rewind functionality to be completely isolated from non-rewind code paths.

gled4er replied:
Thank you for the comment. 

I think this is another example where method name can be improved. That is why I decided to change this method name to `QueryHistoryAsync`. The source code of the method  has nothing specific to the Rewind logic but it just query the **History** table based on a query expression continuously until it can retrieve items. I decided to use it because it is handy that we can pass query condition to it and I did not want to add another method that will do exactly the same.

That said, if you prefer that the Rewind logic works with completely independent from the rest of the logic methods, I can add dedicated method that will be used instead. 

Please let me know if change is required.

Thank you!

gled4er commented:
Hello @cgillum,

Thank you for the great feedback! 

I will work on fixing the issues.

Thank you!

gled4er commented:
Hello @cgillum,

Thank you for your comment about `GetOrchestrationOutputAsync` and `GetOrchestrationInputAsync` methods. I am also worried that their current implementation will introduce performance issues. I would suggest we return large message blob names to the **Instances** table so we can serve all the `GetStatus` calls only from the Instances table without querying the **History** table. 

What do you think?

Thank you!

gled4er commented:
Hello @cgillum,

I forgot to mention that I worked on your feedback from the last review iteration.

Please review when you have time and let me know what should be our approach with the methods you highlighted above.

Thank you!

Kanio

gled4er commented on src/DurableTask.AzureStorage/MessageManager.cs:
```diff
@@ -195,36 +193,39 @@ internal MessageFormatFlags GetMessageFormatFlags(MessageData messageData)
             {
                 BlobResultSegment segment = await instnaceDirectory.ListBlobsSegmentedAsync(blobContinuationToken);
-                storageRequests++;
+                stats.StorageRequests.Increment();
```
Hello @cgillum,

Thank you for the comment. The reason I am not suing this. here is because I decided to pass `AzureStorageOrchestrationServiceStats stats` as parameter for `DeleteLargeMessageBlobs`. I was thinking first to add it as dependency via the `MessageManager` constructor but unfortunately `AzureStorageOrchestrationServiceStats` is less visible than `MessageManager`. I am not sure why `MessageManager` is public class.  

Another option was to move `DeleteLargeMessageBlobs` in `AzureTableTrackingStore` but I think logically the method is more related to the exchanged queue messages. Moreover, in `AzureTableTrackingStore` we don't have `CloudBlobContainer` dependency. 

Please let me know if another approach will be better.

Thank you!

gled4er commented on src/DurableTask.AzureStorage/MessageManager.cs:
```diff
@@ -233,14 +234,12 @@ internal async Task<int> DeleteLargeMessageBlobs(string instanceId)
 
                 await Task.WhenAll(blobForDeletionTaskList);
-                storageRequests += blobForDeletionTaskList.Count;
+                stats.StorageRequests.Increment(blobForDeletionTaskList.Count);
```
Hello @cgillum,

Thank you for the comment.

Please check my reply for your previous comment.

Thank you!

gled4er commented:
Hello @cgillum,

Thank you for the comment for `GetOrchestrationInputAsync` and `GetOrchestrationOutputAsync`. I will be happy to revert the changes but I think we will introduce a bug in the `GetStateAsync` methods. Basically the `OrchestrationState.Input` and `OrchestrationState.Output` will be have the value of the default message I set when we use large message blobs - `Too large to display. Please check History table for the actual data.` instead of the real input and output value.

Moreover, `GetOrchestrationInputAsync` and `GetOrchestrationOutputAsync` depends on `InputBlobName` and `OutputBlobName` and in order to provide the actual input and output data we will need to get the `InputBlobName` and `OutputBlobName` from the **History** table. Below is the previous code for `GetOrchestrationOutputAsync` which is very similar to `GetOrchestrationInputAsync`:

```csharp
async Task<string> GetOrchestrationOutputAsync(OrchestrationInstanceStatus orchestrationInstanceStatus)
        {
            if (string.IsNullOrEmpty(orchestrationInstanceStatus.OutputBlobName))

            {
                return orchestrationInstanceStatus.Output;
            }

            return await this.messageManager.DownloadAndDecompressAsBytesAsync(orchestrationInstanceStatus.OutputBlobName);
        }
```

Please let me know what should be the best approach and I will implement it quickly.

Thank you!

cgillum commented:
Thanks for the explanation. I think it's fine to not include the actual blob payload in the history query response. Ideally we simply return a link to the blob instead of the blob itself. I'm thinking of making that change separately.

gled4er commented:
Hello @cgillum,

Thank you for the clarification.

I will work on the required changes.

Thank you!

gled4er commented:
Hello @cgillum,

May I ask you to take one more look at the latest changes?

I think we made another great optimization that eliminated code and will improve even more the performance of the `GetStateAsync` methods.

Please let me know what should be fixed next.

Thank you!

gled4er commented:
Hello @cgillum,

I run all the storage tests individually and I fixed the once that were failing.

Please take a look when you have time and let me know if I need to fix something.

Thank you!

gled4er commented:
Hello @cgillum,

Thank you very much for your great support and patience until we found the best approach!

It was great to learn so much while working on this feature!

Thank you!

gled4er commented:
Hello @cgillum,

Please let me know when you publish the code via myget and I will prepare quickly the PR for Durable Functions. I can do it later today if this is ok for you.

Thank you!

cgillum commented:
Actually, I'm looking more closely at one of the new tests and the behavior I'm seeing is not what I expected. For example, in the `PurgeInstanceHistoryForTimePeriodDeleteAll` test I see we're running a simple `Echo` orchestration. In theory, there should be exactly 4 blobs:

1. Orchestration input
2. Activity input
3. Activity output
4. Orchestration output

However, the test expects **6** blobs, which does not seem right. I'm guessing that this is because the Instances table has an input and output column, and we're duplicating the blobs there. However, this is quite wasteful since uploading large blobs is expensive in terms of CPU (compression) and I/O (uploading to storage). I would like to understand this better to see if we can reduce this down to four. I'm going to continue investigating.

gled4er commented:
Hello @cgillum,

After debugging the code I see we have 3 blob create calls to [MessageManager.SerializeMessageDataAsync](https://github.com/Azure/durabletask/blob/39e80649b8038889bc9fa5ed2254b3da05662eb2/src/DurableTask.AzureStorage/MessageManager.cs#L38) and 3 blob create calls to[AzureTableTrackingStore.CompressLargeMessageAsync](https://github.com/Azure/durabletask/blob/39e80649b8038889bc9fa5ed2254b3da05662eb2/src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs#L923).

The first 3 calls are direct results of queue message creation and the other 3 are for serializing Azure Table Entities that are happening in AzureTableTrackingStore. 

I will let you know if I find the problem.

Thank you!

gled4er commented:
Hello @cgillum,

I investigated further and the 6 blobs are result because of the following method calls:

` MessageManager.SerializeMessageDataAsync`: 
1. For Control Queue message creation 
2. For Worker Queue message creation
3. For Control Queue message creation

`CompressLargeMessageAsync`: 
1. Call coming from `AzureTableTrackingStore.UpdateStateAsync` for `ExecutionStarted` event 
2. Call coming from `AzureTableTrackingStore.UpdateStateAsync` for `TaskCompleted` event
3. Call coming from `AzureTableTrackingStore.UpdateStateAsync` for `ExecutionCompleted` event

In my opinion it makes sense but maybe I am missing something.

Thank you!

gled4er commented:
Hello @cgillum,

I wanted to follow up on this PR. 

Let me know if I can help with something so we can proceed to the Durable Functions PR.

Thank you!

cgillum commented:
It's better to consider what should be happening by ignoring what the code currently does. I still think there should only be four blobs created in this scenario as I outlined previously. That said, there is no reason to block your PR on this if we think the behavior existed before your changes.

I went ahead and uploaded a newer version of [DurableTask.AzureStorage v1.3.2-rc](https://www.myget.org/feed/azure-appservice/package/nuget/Microsoft.Azure.DurableTask.AzureStorage) to myget.org. Please clear your local nuget cache and you should be able to get your changes in the next nuget restore.

gled4er commented:
Hello @cgillum,

Thank you very much for enabling the code via myget!
